File: frost-taproot/src/context.rs
// Mirrors ref/frost/src/lib/context.ts

use crate::commit::{get_group_binders, get_group_prefix, get_group_pubnonce, get_nonce_ids};
use crate::ecc::group::serialize_element;
use crate::ecc::state::get_point_state;
use crate::ecc::util::lift_x;
use crate::helpers::get_challenge;
use crate::types::{GroupCommitContext, GroupKeyContext, GroupSigningCtx, PublicNonce};
use crate::Error;

/// Build the group key context, applying optional tweaks.
/// Mirrors `get_group_key_context` in the TS implementation.
pub fn get_group_key_context(pubkey: &[u8], tweaks: &[[u8; 32]]) -> Result<GroupKeyContext, Error> {
    let int_pt = lift_x(pubkey)?;
    let int_pk_bytes = serialize_element(&int_pt);

    let group_pt = get_point_state(int_pt, tweaks)?;
    let group_pk = serialize_element(&group_pt.point);

    Ok(GroupKeyContext {
        group_pt,
        group_pk,
        int_pt: Some(int_pt),
        int_pk: Some(int_pk_bytes),
        tweak: None,
    })
}

/// Build the commit context from the key context, nonces, and message.
/// Mirrors `get_group_commit_context` in the TS implementation.
pub fn get_group_commit_context(
    key_ctx: &GroupKeyContext,
    pnonces: &[PublicNonce],
    message: &[u8],
) -> Result<GroupCommitContext, Error> {
    let bind_prefix = get_group_prefix(pnonces, &key_ctx.group_pk, message);
    let bind_factors = get_group_binders(pnonces, &bind_prefix);
    let group_pn = get_group_pubnonce(pnonces, &bind_factors)?;
    let indexes = get_nonce_ids(pnonces);
    let challenge = get_challenge(&group_pn, &key_ctx.group_pk, message)?;

    Ok(GroupCommitContext {
        bind_factors,
        bind_prefix,
        challenge,
        group_pn,
        indexes,
        message: message.to_vec(),
        pnonces: pnonces.to_vec(),
    })
}

/// Build the full signing context.
/// Mirrors `get_group_signing_ctx` in the TS implementation.
pub fn get_group_signing_ctx(
    group_pk: &[u8],
    pnonces: &[PublicNonce],
    message: &[u8],
    tweaks: &[[u8; 32]],
) -> Result<GroupSigningCtx, Error> {
    let key_ctx = get_group_key_context(group_pk, tweaks)?;
    let com_ctx = get_group_commit_context(&key_ctx, pnonces, message)?;

    Ok(GroupSigningCtx {
        group_pt: key_ctx.group_pt,
        group_pk: key_ctx.group_pk,
        int_pt: key_ctx.int_pt,
        int_pk: key_ctx.int_pk,
        tweak: key_ctx.tweak,
        bind_factors: com_ctx.bind_factors,
        bind_prefix: com_ctx.bind_prefix,
        challenge: com_ctx.challenge,
        group_pn: com_ctx.group_pn,
        indexes: com_ctx.indexes,
        message: com_ctx.message,
        pnonces: com_ctx.pnonces,
    })
}

File: frost-taproot-go/context/context.go
// Package context provides signing context creation.
package context

import (
	"math/big"

	"github.com/frost-taproot/frost-taproot-go/commit"
	"github.com/frost-taproot/frost-taproot-go/ecc"
	"github.com/frost-taproot/frost-taproot-go/helpers"
	"github.com/frost-taproot/frost-taproot-go/types"
)

// GetPointState computes the accumulative parity state for a point with tweaks.
func GetPointState(element *ecc.Point, tweaks [][32]byte) (types.PointState, error) {
	pos := big.NewInt(1)
	neg := ecc.ScalarNeg(big.NewInt(1))

	point := &ecc.Point{X: new(big.Int).Set(element.X), Y: new(big.Int).Set(element.Y)}
	var parity *big.Int
	state := pos
	tweak := big.NewInt(0)

	for _, tBytes := range tweaks {
		t := ecc.ScalarFromBytes(tBytes)
		tg := ecc.ScalarBaseMulti(t)

		if ecc.HasEvenY(point) {
			parity = pos
		} else {
			parity = neg
		}

		if parity.Cmp(neg) == 0 {
			point = ecc.NegatePoint(point)
		}

		var err error
		point, err = ecc.ElementAdd(point, tg)
		if err != nil {
			return types.PointState{}, err
		}

		state = ecc.ScalarMul(state, parity)
		tweak = ecc.ModN(new(big.Int).Add(t, ecc.ScalarMul(parity, tweak)))
	}

	if ecc.HasEvenY(point) {
		parity = pos
	} else {
		parity = neg
	}

	return types.PointState{
		Parity: ecc.ScalarToBytes(parity),
		Point:  ecc.SerializePoint(point),
		State:  ecc.ScalarToBytes(state),
		Tweak:  ecc.ScalarToBytes(tweak),
	}, nil
}

// GetGroupKeyContext builds the group key context with optional tweaks.
func GetGroupKeyContext(pubkey []byte, tweaks [][32]byte) (types.GroupKeyContext, error) {
	intPt, err := ecc.LiftX(pubkey)
	if err != nil {
		return types.GroupKeyContext{}, err
	}
	intPkBytes := ecc.SerializePoint(intPt)

	groupPt, err := GetPointState(intPt, tweaks)
	if err != nil {
		return types.GroupKeyContext{}, err
	}

	return types.GroupKeyContext{
		GroupPt: groupPt,
		GroupPk: groupPt.Point,
		IntPt:   &intPkBytes,
		IntPk:   &intPkBytes,
		Tweak:   nil,
	}, nil
}

// GetGroupCommitContext builds the commit context from nonces and message.
func GetGroupCommitContext(keyCtx *types.GroupKeyContext, pnonces []types.PublicNonce, message []byte) (types.GroupCommitContext, error) {
	bindPrefix := commit.GetGroupPrefix(pnonces, keyCtx.GroupPk, message)
	bindFactors := commit.GetGroupBinders(pnonces, bindPrefix)
	groupPn, err := commit.GetGroupPubnonce(pnonces, bindFactors)
	if err != nil {
		return types.GroupCommitContext{}, err
	}
	indexes := commit.GetNonceIDs(pnonces)

	challenge, err := helpers.GetChallenge(groupPn[:], keyCtx.GroupPk[:], message)
	if err != nil {
		return types.GroupCommitContext{}, err
	}

	return types.GroupCommitContext{
		BindFactors: bindFactors,
		BindPrefix:  bindPrefix,
		Challenge:   ecc.ScalarToBytes(challenge),
		GroupPn:     groupPn,
		Indexes:     indexes,
		Message:     append([]byte(nil), message...),
		Pnonces:     append([]types.PublicNonce(nil), pnonces...),
	}, nil
}

// GetGroupSigningCtx builds the full signing context.
func GetGroupSigningCtx(groupPk []byte, pnonces []types.PublicNonce, message []byte, tweaks [][32]byte) (types.GroupSigningCtx, error) {
	keyCtx, err := GetGroupKeyContext(groupPk, tweaks)
	if err != nil {
		return types.GroupSigningCtx{}, err
	}

	comCtx, err := GetGroupCommitContext(&keyCtx, pnonces, message)
	if err != nil {
		return types.GroupSigningCtx{}, err
	}

	return types.GroupSigningCtx{
		GroupPt:     keyCtx.GroupPt,
		GroupPk:     keyCtx.GroupPk,
		IntPt:       keyCtx.IntPt,
		IntPk:       keyCtx.IntPk,
		Tweak:       keyCtx.Tweak,
		BindFactors: comCtx.BindFactors,
		BindPrefix:  comCtx.BindPrefix,
		Challenge:   comCtx.Challenge,
		GroupPn:     comCtx.GroupPn,
		Indexes:     comCtx.Indexes,
		Message:     comCtx.Message,
		Pnonces:     comCtx.Pnonces,
	}, nil
}
